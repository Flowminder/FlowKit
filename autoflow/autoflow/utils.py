# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

"""
Utility functions.
"""

import collections
import json
import nbconvert
import nbformat
import networkx as nx
import pendulum
from contextlib import contextmanager
from get_secret_or_env_var import getenv
from hashlib import md5
from pathlib import Path
from sh import asciidoctor_pdf
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from tempfile import TemporaryDirectory
from typing import Any, Dict, Iterator, Optional, OrderedDict, Set, Tuple, Union


def get_output_filename(input_filename: str, tag: str = "") -> str:
    """
    Given an input filename, construct an output filename with the same extension,
    with a timestamp (and optionally a tag) added to the stem.

    Parameters
    ----------
    input_filename : str
        Input filename
    tag : str, optional
        A tag to append to the file stem
    
    Returns
    -------
    str
        Output filename
    """
    input_filename = Path(input_filename)
    now_string = pendulum.now("utc").format("YYYYMMDDTHHmmss[Z]")
    tag = f"__{tag}" if tag != "" else tag
    return f"{input_filename.stem}{tag}__{now_string}{input_filename.suffix}"


def get_params_hash(parameters: Dict[str, Any]) -> str:
    """
    Generate a md5 hash string from a dictionary of parameters.
    The dictionary is dumped to a json string before hashing.

    Parameters
    ----------
    parameters : dict
        Dictionary of parameters to use for creating the hash.
    
    Returns
    -------
    str
        md5 hash of the parameters dict
    """
    return md5(
        json.dumps(dict(parameters), sort_keys=True, default=str).encode()
    ).hexdigest()


def get_session(db_uri: str) -> "sqlalchemy.orm.session.Session":
    """
    Create a sqlalchemy session.

    Parameters
    ----------
    db_uri : str
        Database URI
    
    Returns
    -------
    Session
        A sqlalchemy session
    """
    # TODO: This seems like the wrong place to be reading a secret / env var,
    # but we can't put a docker secret in the prefect config.
    full_db_uri = db_uri.format(getenv("AUTOFLOW_DB_PASSWORD", ""))
    engine = create_engine(full_db_uri)
    return sessionmaker(bind=engine)()


@contextmanager
def session_scope(db_uri: str) -> Iterator["sqlalchemy.orm.session.Session"]:
    """
    Provide a scoped sqlalchemy session for interacting with the database.

    Parameters
    ----------
    db_uri : str
        Database URI
    
    Yields
    ------
    Session
        A sqlalchemy session
    """
    session = get_session(db_uri)
    try:
        yield session
        session.commit()
    except:
        session.rollback()
        raise
    finally:
        session.close()


def make_json_serialisable(obj: Any) -> Union[dict, list, str, int, float, bool, None]:
    """
    Helper function to convert an object's type so that it can be serialised using
    the default json serialiser. Non-serialisable types are converted to strings.

    Parameters
    ----------
    obj : any
        Object to be converted.
    
    Returns
    -------
    dict, list, str, int, float, bool or None
        The same object after dumping to json (converting to str if necessary) and loading
    """
    return json.loads(json.dumps(obj, default=str))


def get_additional_parameter_names_for_notebooks(
    notebooks: Dict[str, Dict[str, Any]],
    reserved_parameter_names: Optional[Set[str]] = None,
) -> Set[str]:
    """
    Extract parameter names from a list of notebook task specifications.

    Parameters
    ----------
    notebooks : dict
        Dictionary of dictionaries describing notebook tasks.
    reserved_parameter_names : set of str, optional
        Names of parameters used within workflow.
    
    Returns
    -------
    set of str
        Names of parameters used by notebooks which must be passed to a workflow.
    """
    # Parameters requested in notebooks
    notebook_parameter_names = set.union(
        *[set(notebook["parameters"].values()) for notebook in notebooks.values()]
    )
    # Parameters available to notebooks, which are not passed externally as flow parameters
    internal_parameter_names = (reserved_parameter_names or set()).union(
        notebooks.keys()
    )
    # Additional parameters required for notebooks
    additional_parameter_names_for_notebooks = notebook_parameter_names.difference(
        internal_parameter_names
    )
    return additional_parameter_names_for_notebooks


def sort_notebooks(
    notebooks: Dict[str, Dict[str, Any]]
) -> OrderedDict[str, Dict[str, Any]]:
    """
    Perform a topological sort on a dictionary of notebook task specifications.

    Parameters
    ----------
    notebooks : dict
        Dictionary of dictionaries describing notebook tasks.
    
    Returns
    -------
    OrderedDict
        Ordered dict of notebook task dicts, ordered so that no notebook depends on another
        notebook that comes after it.
    
    Raises
    ------
    ValueError
        If the notebook specifications contain circular dependencies.
    """
    notebooks_graph = nx.DiGraph(
        {
            key: [
                value
                for value in notebooks[key]["parameters"].values()
                if value in notebooks
            ]
            for key in notebooks
        }
    ).reverse()
    try:
        sorted_notebook_keys = list(nx.topological_sort(notebooks_graph))
    except nx.NetworkXUnfeasible:
        raise ValueError("Notebook specifications contain circular dependencies.")

    return collections.OrderedDict(
        (key, notebooks[key]) for key in sorted_notebook_keys
    )


def notebook_to_asciidoc(
    notebook_path: str, asciidoc_template_path: Optional[str] = None
) -> Tuple[str, "nbconvert.exporters.exporter.ResourcesDict"]:
    """
    Convert a Jupyter notebook to ASCIIDoc, using nbconvert.

    Parameters
    ----------
    notebook_path : str
        Path to notebook
    asciidoc_template_path : str, optional
        Path to a template file to use for the conversion.
        If not provided, the default nbconvert asciidoc template will be used.

    Returns
    -------
    str
        The resulting converted notebook.
    ResourcesDict
        Dictionary of resources used prior to and during the conversion process.
    """
    with open(notebook_path) as nb_file:
        nb_read = nbformat.read(nb_file, as_version=4)

    if asciidoc_template_path is None:
        exporter = nbconvert.ASCIIDocExporter()
    else:
        exporter = nbconvert.ASCIIDocExporter(template_file=asciidoc_template_path)
    body, resources = exporter.from_notebook_node(nb_read)

    return body, resources


def asciidoc_to_pdf(
    body: str, resources: "nbconvert.exporters.exporter.ResourcesDict", output_path: str
) -> None:
    """
    Convert an ASCIIDoc document to PDF, using asciidoctor-pdf.

    Parameters
    ----------
    body : str
        Content of the ASCIIDoc document
    resources : ResourcesDict
        Dictionary of resources, as returned from nbconvert exporter
    """
    with TemporaryDirectory() as tmpdir:
        with open(f"{tmpdir}/tmp.asciidoc", "w") as f_tmp:
            f_tmp.write(body)
        for fname, content in resources["outputs"].items():
            with open(f"{tmpdir}/{fname}", "wb") as fout:
                fout.write(content)
        asciidoctor_pdf(f"{tmpdir}/tmp.asciidoc", o=output_path)
