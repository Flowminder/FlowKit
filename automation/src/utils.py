from pathlib import Path

def get_output_filename(input_path, output_dir, label):
    input_path = Path(input_path)
    return f"{output_dir}/{input_path.stem}_{label}{input_path.suffix}"