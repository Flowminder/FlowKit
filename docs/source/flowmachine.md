# FlowMachine

FlowMachine is a Python toolkit for the analysis of CDR data. It is essentially a python wrapper for postgres SQL queries, including geographical queries with postgis. All of the classes in flowmachine define an SQL query under the hood.

## Documentation

Documentation for FlowMachine can be found [here](../developer/api/flowmachine).
