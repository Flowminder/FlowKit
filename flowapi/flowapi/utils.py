# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.


from quart import request
from quart.exceptions import HTTPException


async def get_query_parameters_from_flowmachine(*, query_id) -> dict:
    """
    Get the parameters of a query from flowmachine.

    Parameters
    ----------
    query_id : str
        ID of the query to get params for

    Returns
    -------
    dict
        Dictionary containing the query's original parameters

    Raises
    ------
    HTTPException
        404 if the query id is not known.

    """
    request.socket.send_json(
        {
            "request_id": request.request_id,
            "action": "get_query_params",
            "params": {"query_id": query_id},
        }
    )
    reply = await request.socket.recv_json()
    if reply["status"] == "error":
        raise HTTPException(
            description=f"Unknown query ID '{query_id}'",
            name="Query ID not found",
            status_code=404,
        )
    return reply["payload"]["query_params"]
