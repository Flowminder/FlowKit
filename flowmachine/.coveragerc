[run]
branch = True

[report]
fail_under = 90

omit =
    */python?.?/*
    */venv/*
    */site-packages/nose/*
    */tests/*
    */_version.py
    */versioneer.py
    *__init__*
    flowmachine/core/server/server.py

exclude_lines =
    if __name__ == '__main__':

show_missing = True
