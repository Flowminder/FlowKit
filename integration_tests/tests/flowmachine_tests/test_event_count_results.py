# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

from flowmachine.utils import pretty_sql

from approvaltests.approvals import verify
from flowmachine.core import CustomQuery
from flowmachine.features import EventCount


def test_event_count_1_sql(diff_reporter):
    """
    Event count query with non-default parameters returns the expected data.
    """
    ec = EventCount(start="2016-01-02", stop="2016-01-04", tables=None)
    sql = pretty_sql(ec.get_query())
    verify(sql, diff_reporter)


def test_event_count_1_df(get_dataframe, diff_reporter):
    """
    Event count query with non-default parameters returns the expected data.
    """
    ec = EventCount(start="2016-01-02", stop="2016-01-04", tables=None)
    df = get_dataframe(ec)
    verify(df.to_csv(), diff_reporter)
