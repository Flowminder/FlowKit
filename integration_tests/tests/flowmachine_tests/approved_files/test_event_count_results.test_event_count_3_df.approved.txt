,subscriber,value
0,2ZdMowMXoyMByY07,3
1,MobnrVMDK24wPRzB,5
2,0Ze1l70j0LNgyY4w,6
3,Nnlqka1oevEMvVrm,3
4,4dqenN2oQZExwEK2,6
5,gPZ7jbqlnAXR3JG5,8
6,G0VxgAwoOPAM685Q,6
7,XWRldepdNlegN3Zo,3
8,0NVaL8k1nwRwOmlg,3
9,rJeX25p8mn5Qwqmd,10
10,NzlPGkByB5jqBJ1v,5
11,NOXLME8YWo09J6oY,7
12,RGkzrL0JYKb2gPqW,6
13,A42Xl6GzjBqMODyz,8
14,QWZzERBdXVM0GKXd,3
15,bKZLwjrMQG7z468y,4
16,gZPKMqedeKrQ72pn,3
17,l81xEJB2XdJWNedP,8
18,37J9rKydzJ0mvo0z,5
19,GJP3DWdGyb4QBnyo,5
20,self_caller,2
21,qPorDyvDD8RmgnL0,5
22,Vqjp2yxLk8yebXmP,3
23,oQOm1YkDxljp3AVL,6
24,A8Y0ylLjaDDQwORD,3
25,pB8g1y7jQEZO734d,3
26,xN4XQEa6JZJvkeAj,4
27,wrmEaJ9prNvyV4zL,9
28,eJ6vRwmbMEv5E7gq,2
29,49x6bnGG8mnWMy87,7
30,E0LZAa7AyNd34Djq,6
31,DzpZJ2EaVQo2X5vM,7
32,6zm4b7veY2ljEQ95,6
33,poXqlQOnjlJ5EZMj,4
34,Q3xOLdlXJ4dMB7R1,6
35,qEzo57xNeXjLBDeX,11
36,E1n7JoqxPBjvR5Ve,6
37,NA7D19MmEm9gkK5R,2
38,BNxWrJ9mqMJ83Mzk,7
39,ZQbqE6Kw4d6WRnm2,3
40,Q7lRVD4YY1xawymG,6
41,8dpPLR15XwR7jQyN,11
42,KyonmeOzDkEN40qP,10
43,NG1km5NzBg5JD8nj,2
44,N5lAaO2KM31PRzp6,8
45,pd0bV8W1z4kJzgv5,5
46,81B6q0K8k325XWmn,7
47,B7blVrVn0br1zP3x,5
48,KAygLZgx8oAeJz8Y,3
49,kjGXLy9lWnZ4V6J7,4
50,8b6qy2aDVmjlaogE,2
51,P7kMZxeEOzxr2Wg1,7
52,Ln4ly8DYR6wzWDMo,6
53,M60LdelQ6ROXRJ3P,2
54,Rzx9WE1QRqdEX2Gp,5
55,p6alg5rYl14BG2nL,3
56,bjY7mMXxE3zMoelO,2
57,O8a7ADMbbmv0QkbV,5
58,zj9KZblppy6byEDg,5
59,AMjGX58pWv5Ql2o6,2
60,gwAynWXp4eWvxGP7,7
61,8gQ13KpqedDJO4aE,6
62,X2GyrjgBPNJpveaZ,4
63,yYbzQ46yEx93D7je,4
64,gWPl7QBD8VnEyX8K,7
65,x8o7D2Ax8Lg5Y0MB,8
66,gKkzQLa0yZ6Mvonm,3
67,aK9GDX8nozb2RB67,3
68,yObw75JkAZ0vKRlz,1
69,ZYPxqVGLzlQy6l7n,5
70,D6b8NwVBmmw5JzA1,2
71,8rxEQOL3ePZdAe1z,1
72,O2oGlMjNX7MpwV43,4
73,p8oPjnDRq0M2wJW0,7
74,GXd06jGNEamgBwr3,7
75,dw0xVe5rw3gy14EL,6
76,5KaRV2B36vpNoybG,8
77,PeJKXdZRoadyEMaj,2
78,JO3QBk9PmjoMmge6,3
79,nGXqgLRgBKlRd79W,4
80,V7MBRewnwQGE91gY,9
81,Wm2PbGlxP7GAzvVn,6
82,BebaVL8XZzLn18y2,4
83,zGWn8opVmOQAD6xY,8
84,1jwYL3Nl1Y46lNeQ,4
85,oanJmEDrLxpbVPzM,5
86,LnKG6xoRJ0X1wm7Y,2
87,OBwVW2WZNdMrEv7Q,7
88,NBpb54xVMEw1Yqza,6
89,VkzMxYjv7mYn53oK,7
90,5Q0lndPxZYdZGMO8,6
91,Vqy0OnN9P6lYgAze,6
92,BGXbdrXG9J2gVZoA,1
93,oqNR8gkbv6e4K97z,5
94,o1Rwj2XPYo2E05KP,5
95,peMgdn594XnNO4bK,2
96,kw0rRZVr2vZB2zWl,2
97,YGE2AwGPPEa86kq9,7
98,2PjYkGv6ARGqQ67p,4
99,4ZVwldpn4QXDN0OY,6
100,YNv2EgDJxxAoy0Gr,9
101,DWV9mxjjvLxwnXyb,8
102,87mobXjNNDVX51ML,4
103,NvzxDwNZo0wleBg5,6
104,Q5wy089DDNQG4AEX,4
105,7nWrYGEd029zJ02M,4
106,nJLe9jzo96KVAXN7,2
107,P5Bwzx72E80Dno21,3
108,Bn5kZrQ2WgEy14zN,6
109,ABV1RygG9oyXoJ8j,3
110,P8RpJMqGpexyOl2o,1
111,ka54Xb0N4x9LqPnE,1
112,vKVLDx8koQWZ2ez0,4
113,NYdrjyrdMdZDmXok,3
114,6v9YyNx3r7zb3kO8,8
115,X809mynDXN621PoN,2
116,nX89Y35pYvqAVvpD,7
117,OdM7np8LYEp1mkvP,4
118,2Dq97XmPqvL6noGk,6
119,ed57vNjYWoLxOBMl,2
120,jW09JkYPjv6zmKP8,4
121,5zgW0QZlRLvwrLJo,5
122,gWXN3Z5veAr162kw,4
123,gKqw0y4RWwQApPrm,11
124,6ALdEroGZZrjqbgl,4
125,X12gvGBdjp7E3Pbj,5
126,0ayZGYEQrqYlKw6g,7
127,YK6z2lXzg7w57Vap,5
128,4ZpLvnNZArnNae9R,5
129,1vGR8kp342yxEpwY,8
130,xkZb5E55LYE10wa4,3
131,0Gl95NRLjW2aw8pW,5
132,W8eEBvV6P8Jv01XZ,8
133,JNK7mk5G1Dy6M2Ya,5
134,l2RN3b7WRl2b1V8P,4
135,DAlqeZENbeOn2vBw,6
136,4oLKbnxm3vXqjMVx,4
137,amW6DwnDmbZ50xkO,4
138,jWXpmeg35Ye3ByoJ,4
139,5vwqabdVV1lZyLBE,9
140,AZj6MqBAryVyNRDo,7
141,YxDnjbQ4E9pVK4A6,8
142,83Kywz7PZqjMPJpR,4
143,m51bzWYMxKWLJy2V,4
144,4pQo67v0PWyLdYKO,7
145,w1N59Jo07vbgrKae,7
146,wynMvOdqX6Eo9PmR,5
147,MLo57YoM963gRQe1,7
148,QWAjL9Nbyx95moq7,4
149,JN6WPrJPeKV1lx3q,3
150,Rv8rPlM4wp54q7QZ,4
151,YR920VxJq5PNQBAG,3
152,QK78nqVO0o21VGB0,7
153,gWwV6lpVON7ryzDe,4
154,YeqxBmVgL1EnONv8,4
155,8br1gO32xWXxjY0R,3
156,ZQG8glazmxYa1K62,5
157,KD1ro51Yo6o5vg6n,5
158,EkpjZe5z37W70QKA,8
159,8lo9EgjnyjgKO7vL,2
160,JZoaw2jzvK2QMKYX,3
161,By4j6PKdB4NGMpxr,5
162,oq936AVWZRzYWPrZ,4
163,KmdyQlgVyQjDWvJN,2
164,lDv0jon4AyqL4byn,4
165,Ea2DjkNvAvkp7OMe,9
166,M9YvRrkbpeZ052yj,8
167,BVYqp0ryO1oj1gRo,5
168,Y8lvgqN0n75GQwk7,7
169,V4DJ7AaxLLOyWEZ3,6
170,vJrRPQRm1lEma2KO,4
171,wMbWOpJOk0GmZvLD,6
172,W2VRAqvL4OM1xZEO,7
173,VgzRWn82MdlbXkyO,5
174,VGYEqRGVxVQl7JZO,5
175,gkBLe0mN5j3qmRpX,4
176,jWlyLwbGdvKV35Mm,3
177,M0pdYQAADkQXP9Vy,5
178,dWE82PWRRkX5wr0b,6
179,lkPBK9BmJMpvMWnb,5
180,Pq0xpgkyDrY6kmDy,4
181,aroL1Ar5K1jK9G26,4
182,a0ZQeXdY48Wg5bkw,6
183,v3kQyzZNRNAz5lbZ,4
184,0zqZPERbraE8NM4W,2
185,dgz4WNZOAm5bmJ9M,6
186,vBR2gwl2B0nDJob5,5
187,w2YgmOPlr2O5x3VJ,5
188,G2DQzae1qOa48jK9,4
189,ApNYxdeQlEdm76ga,8
190,Bar6kR1QXKRlYn1J,3
191,B4JG213EnO0bwXAN,6
192,ya9DBmDpJNm25lP1,3
193,mqP6d9k5nR21oaeb,2
194,EreGoBpxJOBNl392,6
195,8e4zNGWqVArxE2Zo,5
196,RB4gWNwl4D8exP7l,4
197,M9GgmAJX2g75aK2N,10
198,7dePlr8JjqRQ9vGz,9
199,ZM3zYAPqx95Rw15J,4
200,lMPW3K768jbgyRvo,6
201,KM4YzL5aPRL59dRy,3
202,EdWZO2dVLNRDMPlV,3
203,rG5oYMVaBj42Oa6L,9
204,X47N9O28wGd6Mneo,3
205,q4rpo9aoJR3BbvWl,9
206,7XebRKr35JMJnq8A,1
207,WkYAGXV6A1XOxg8o,7
208,dM4aLP8N97eYABwR,3
209,mkqQ4NPBPQLapbeg,3
210,5oDLGWg9lBeQRP9d,4
211,3x5BKX29EzQVDbJy,4
212,2vmOlAENnxpPM1xX,3
213,LjDxeZEREElG7m0r,8
214,pxgnQqP14BVkLM6o,6
215,gQ0eDL57DgxKwqn8,6
216,WdBX5pVBE2pynxPQ,2
217,QEoRM9vlkV18N4ZY,4
218,oL13aVowaNpVk2np,1
219,gz3o1Zr6yJZnOYr9,2
220,lRjYnM34X5AXBGrz,3
221,09NrjaNNvDanD8pk,7
222,kmR2eMYzRbAW8r4Q,2
223,E9nlO1dMyVAWr60X,3
224,148ZaRZe54wPGQ9r,4
225,gPDW4zr1YmVGqEAO,5
226,oX6kEmV5yyb0xq3M,1
227,nAELXw76mverkmqZ,4
228,QrAlXqDbXDkNJe3E,4
229,J0Yyqw2rkVEwpMG2,4
230,7JyRXDKdd4gnwq0d,3
231,7lNP0mDOAK3xKWv4,4
232,GNmpb80ZawxQ7WOV,10
233,pdrXzZj2DkZLJvbk,5
234,a76Ajyb9dmEYNd8L,4
235,KXVqP6JyVDGzQa3b,5
236,4587JeOdRBEW96K1,4
237,LqgwYrR7ZZp5xGK1,4
238,kq7DLQar1nw2glYN,3
239,NejWvxnkxjlXpJgQ,7
240,4PWQE7lNA25NYV6p,4
241,PWlZ8Q5zjYkGJzXg,6
242,038OVABN11Ak4W5P,5
243,5Kgwy8Gp6DlN3Eq9,5
244,ye8jQ0ovnGd9GlJa,3
245,woOyJvzKBALYj8za,7
246,j6QYNbMJgAwlVORP,3
247,NeRzMVGy2MEvgjDZ,5
248,8bGZXqJXP1Jgr1EM,4
249,6nYkgBqZNo5KG7N4,3
250,oLkvlE7D98R6XK0z,10
251,BKMy1nYEZpnoEA7G,7
252,kYp5Xor8eGQ0DBZL,4
253,RkdLwPWLboQ2WvX4,3
254,RnAv0YjPnW68x9wr,5
255,Ma0Xv1XJxE2JV7r8,6
256,ZEjaJ0lkr19vYWRB,9
257,pD0me6Do0dNE1nYZ,3
258,nB4qp1mVza9bWY0E,1
259,vOGRpPpAArjBQZ9y,5
260,Z89mWDgZrr3qpnlB,5
261,lqOknAJRDNAewM10,7
262,rmb4PG9gx1ZqBE01,6
263,yqw50eNyEwOxNDGL,4
264,EYojLpzGVbzyJQvn,5
265,AgB6KR3Levd9Z1vJ,8
266,m4L326vrwE6elJxQ,6
267,QeREg30jMjAd0Vz4,3
268,Ak4LV2l6yNDrg9qw,3
269,gV57Wm5zvWdklYjz,7
270,MXLPGvyz63An4bE1,5
271,Q4mwVxpBOo7X2lb9,6
272,AwQr1M3w8mMvV3p4,4
273,onayVzV3pWrpW3M8,9
274,GnyZMedmKQ4X78Wa,2
275,Da7ewlw4Ok1yvZmB,2
276,a7wRAnjb2znjOlxK,8
277,o4ya5j92g5GRxdw2,15
278,yqQ6m96rp09dn510,9
279,jwKJorl0yBrZX5N8,6
280,QoD6rWNN03WPbEBg,4
281,1p4MYbA1Y4bZzBQa,6
282,PzQk5poRMA7ERaBe,4
283,2Qw3PGwZJdqaZYbJ,3
284,Aa3kjzJmqZWveXqb,4
285,gxz7rWG7rVWj40vX,5
286,p7N2YgEKM18vwmLe,4
287,z7mQgkVYoz2vr4EG,5
288,4g7a3XELre5dYzr6,4
289,7ra3xZakjEqB1Al5,6
290,APj9roe8jKOwEDZl,7
291,wvMm1EAY0QeEAbN9,2
292,l7xALKEXZqjPaWQY,12
293,wZReL3b1lEQkrm2G,6
294,Egly7ZQvwdopGmYP,5
295,eg5G9xM8aZAmnKp1,6
296,dXogRAnyg1Q9lE3J,8
297,pnwa9PE5MbPB46O0,3
298,9vXy462Ej8V1kpWl,2
299,WKP7OnqzA2rxDA6J,6
300,3GZ48OAlG6mPMBKg,5
301,58nJwlYY8El6VX0E,3
302,aJmlqpLqaa8kn94P,5
303,1QBlwRo4Kd5v3Ogz,6
304,mJ9eZYnvvr5YGW2j,7
305,Kabe7EppYbMEj6O3,7
306,mV1dOAOE0Vq4DLJk,7
307,7nqEVPQ06ZkQ9w3a,3
308,mKyqZPM1AXPAQL08,3
309,OemQ7q2DLZMWnwzB,3
310,KyNxbYLE2Qzj3mPD,2
311,o5V0lPmGoorXnv4q,4
312,nvKNoAmxMvBW4kJr,1
313,0gmvwzMAYbz5We1E,10
314,5jLW0EWeoyg6NQo3,7
315,dr9xNYK006wykgXj,5
316,Mmz20GVDbqjEel51,8
317,y2MZGq0N6370KXoz,2
318,qRlQo7ly3Zg9GVN2,5
319,DELmRj9Vvl346G50,6
320,YMBqRkzbbxGkX3zA,5
321,0W71ObElrz5VkdZw,6
322,2dpYQW4pzMWBxORA,8
323,dlyoVBZ5LR9J7QYW,2
324,bMrnX9xrVNp0lvx2,7
325,Kxj9rpMEA9pMmV8J,6
326,jpXleRQqrLR0aPwN,3
327,QWjwV5Xr3zEeKoDP,8
328,Kxb2JoGX50lzRDEq,5
329,8dXPM6JXj05qwjW0,3
330,qj8gk6qbN8R3DBdO,5
331,5xZ8YkbRqrkv1r4J,4
332,2NPz1Vz1AaKDLoGQ,7
333,8DLbw958mPv3xjVa,9
334,MX0JR5y11e5narO6,3
335,9eDMvX85yLR3dGlE,4
336,4QyYADMYXqPDP79G,5
337,b3EK7zNOVVAqmaOn,5
338,3EgqzplqPYDyGRVK,4
339,6qMK0QbqlAnWl2dN,7
340,Ap3E6Ljl0kWRavy4,2
341,nL9KYGXpz2G5mvDa,6
342,alye4Z3yz5ENvnXY,2
343,4G1xdE6J3gEeVJnv,2
344,mrlW4GBWMO6DABXx,9
345,jNMBy0rVVd0DnGK6,1
346,e6BxY8mAP38GyAQz,7
347,rBwRoXaBP6LM1pzP,8
348,8GkR7zbjLezebY5E,6
349,PGmKQ0PqdLkXA9qE,4
350,B4A8p0WW1Q0xk92Y,8
351,JRZMbq2eOM189Qw4,8
352,N6R3Akx0zWk05oWp,3
353,PR31Yr7agDbJzaZX,7
354,Bwo5pgJzV1ogxyeK,4
355,lVPvMWVBKOW74ZEO,7
356,oX7ZVm05RPr9v0Aq,2
357,W6ljak9BpW0Z7ybg,4
358,pDNARk60aQwO9Qar,7
359,PnxGEed4JWezYdBg,6
360,V1QBpMj0vEwr2PGW,2
361,Gd2bxpQG0xR91VXD,12
362,6NKDrZDNWwayGO9j,5
363,kJRMlxA71OL5om49,1
364,Vqvyp1AGr7m58D4x,4
365,7qKmzkeMbmk5nOa0,1
366,8JeRLGmxlq0k0qZz,7
367,Le5bxBw5VpqZpGQ2,5
368,k8JRVodMrpKQp35q,3
369,ZeloJKzkEwVrLNYO,7
370,jwgy7YympQYMBQVx,2
371,qPAbaED3vDYkjZ0n,4
372,GM421LnWGWDaEPl9,3
373,PJ536E18n1D4Bd1r,2
374,4AqjP2LWpw3r73vB,6
375,l9X0ZO9AGrOg62RM,1
376,8qkMv7moe9b9AowY,4
377,JrYLpWRmryNm0xDP,7
378,xE8QMJxmLGMW6lZ4,6
379,xJwr9glV6emG7b8o,9
380,1d29oEA95KEzAKlW,6
381,NQV3J52PeYgbLm2w,12
382,lGB9advjJYjdbWrN,7
383,PagZ0KRaVNlY58kQ,4
384,9GMx01aQZx13E48l,9
385,ZmPRjkQ74Xeql71V,3
386,N5vEaB5gO5yDxRrl,6
387,8Nerw0xj58Ky5RVz,4
388,Oko9NEkexzJQKReX,5
389,0DB8zw67E9mZAPK2,6
390,QDqWpen1qakOLzd5,7
391,DLajEKM0gNPZGJeA,5
392,D9qXvyM1NbyxpdEm,5
393,a3PrBbGBAXpL8mp0,4
394,vJK0DwMjZEm7gBQV,8
395,nRLeDl77Del5xpbY,2
396,l8KM7myrkvPLXpZx,1
397,xoJnvpa7LJNgyqKe,6
398,5yPVlxl9aDbK2L0r,4
399,2MdDGpm37Dr41nAY,8
400,1NqnrAB9bRd597x2,3
401,QPdr2B94VaEzZgoW,10
402,2yqlaY6LD9Q8xZj1,4
403,wX9ODedAaV5g8v26,6
404,6wGDKd79v0bnMzor,7
405,ZnYwpaZxEjW6Ke7k,4
406,Q0ELoRGG3X9Rg1GA,5
407,AgvE8pa3Bvqezmo6,3
408,XjyWlpdZNZzY3O0N,2
409,Z4pBaoQLj4Gz76KA,5
410,bejv8KjzaA7K0319,1
411,RZkbX2rypqWja9Bv,5
412,XRo0EdKqwD3zZBLr,6
413,GNLM7eW5J5wmlwRa,5
414,LBlWd64rqnMGv7kY,6
415,24ndeQw41nGqlPZ1,2
416,Yd7BeVPEJA0xlJwr,4
417,bvEWVnZdwJ8Lgkm2,4
418,2GJxeNazvlgZbqj6,3
419,ngRw8aKKdpaEmY3Q,4
420,Aqp2DjeEmwjEVZKo,2
421,Q1jMk7qjqXBnwoDR,6
422,2ZRmOxEWBVq546zY,5
423,b9Y3V7deAlgXolRd,6
424,vWwlyp5a4GpdxmBZ,11
425,J7gz8x0KJKn5y3MG,5
426,EqrX8mBXDZPMRKLz,5
427,OzlrV3GEOyn5DW1Y,6
428,qJjd9VmvDrgQMA1O,4
429,Oxv502qrde8KmgVr,6
430,mpqkOe1L1wAaMRgE,2
431,naXZEREAWwR9Nl6x,6
432,nLvm2gVnEdg7lzqX,6
433,yPzO2eJRrzeqnkEV,3
434,oR78lLNzN9LpgqNb,5
435,wyrgMxBzzMxQVj83,6
436,k9J7lEbPxBnVyj3D,5
437,7wmpbZzKRnaR4GvO,4
438,D8AZvkA5NXmkyEeO,1
439,3XKdxqvyNxO2vLD1,5
440,OkNQl0jqMveDx5Rv,5
441,7b4BZo0kwQYRmyJ0,3
442,jERG2g84xGgVyz6K,2
443,g9D5KEK9BQzOa8z0,7
444,wEZaqV15wMVGe1v6,2
445,W6PRYNgyW89eZ8A1,5
446,J4K6gkq7wyqopOnQ,2
447,L4V537alj321eWz6,4
448,M87PjXJVWroXROWw,3
449,Zv4W9eak2QN1M5A7,4
450,x7wGEeYdBGD9aQ5X,1
451,GMKmXnYNgVRlyb6r,4
452,DX2eLoNWReGEPYnK,6
453,lPp65Z8gbEywqOnG,3
454,MZze6BkgGoBv5dEG,4
455,9wkjq8LdrAr84Y6g,3
456,RDqQa5nOVoX07Lk4,4
457,Nbng985Vjoa0Ze1X,8
458,0MQ4RYeKn7lryxGa,7
459,Z51jagNxqkWR2L0o,5
460,DpmEKQaW8X3PdZLo,4
461,3enxoqbomNMB97lW,6
462,KAMQan1jjab4rlgX,9
463,PbN9vQlbEORLmaxd,5
464,4QVGRaVGwydYE86n,7
465,PZVjpGbPvyjgQErM,8
466,oWkryLvyr0Jba1E5,5
467,Bp8QkGlv3E4mxvOP,1
468,jJOA5l9NyPEmVrPK,5
469,jPk9xAzO70rR1bBe,5
470,EapG7nvNNpn1PAdv,5
471,MRwEk8RmvK2Y0P9N,6
472,wgoa5OApGQey32ZM,6
473,3DnLz7g07mkQepaO,6
474,N640d1AGW8pyxWKn,8
475,5N1xXmGoy0mq2v08,4
476,WaekA0LbqwO6rpRG,5
477,r1av4JnM18dYGxAe,4
478,Zy3DkbY7MDd6Er7l,6
479,6NWOjmXPkxmbaw73,6
480,jNQ9L7MjWLGebYM6,5
481,xDQAPjbY0Z5o3Lqg,3
482,nMvpK39bowVXYN9G,6
483,346j7Nq67nvXZR0m,7
484,l6NQwr8qzwZLGy92,5
485,Be937K2jdjKnoRW6,5
486,XQOrADNd8ePa71Ve,4
487,alK9yYKwJ3YgXqJW,8
488,okNGOMXqlJjnRarA,7
489,951zd63qryZAYBy7,3
490,lLVzNDy2BX1aO8wY,11
491,5q2VK37925edBXb8,4
492,zvaOknzKbEVD2eME,9
493,LR6Dn0DNJeGx39g8,5
494,6kj8RD7YQ4kwWlQy,3
495,62QyANzLydvMzVgm,4
496,JAqlnPX1QmZ8pQ5V,5
497,dqEQWNwAYEr4Mk3e,9
498,3o9YZlRpm74nwDXx,8
